sycl/include/sycl/detail/id_queries_fit_in_int.hpp

```diff
@@ -34,46 +34,96 @@ inline namespace _V1 {
 namespace detail {
 
 #if __SYCL_ID_QUERIES_FIT_IN_INT__
-template <typename T> struct NotIntMsg;
+constexpr static const char *Msg =
+    "Provided range and/or offset does not fit in int. Pass "
+    "`-fno-sycl-id-queries-fit-in-int' to remove this limit.";
 
-template <int Dims> struct NotIntMsg<range<Dims>> {
-  constexpr static const char *Msg =
-      "Provided range is out of integer limits. Pass "
-      "`-fno-sycl-id-queries-fit-in-int' to disable range check.";
-};
-
-template <int Dims> struct NotIntMsg<id<Dims>> {
-  constexpr static const char *Msg =
-      "Provided offset is out of integer limits. Pass "
-      "`-fno-sycl-id-queries-fit-in-int' to disable offset check.";
-};
-
-template <typename T, typename ValT>
+template <typename ValT>
 typename std::enable_if_t<std::is_same<ValT, size_t>::value ||
                           std::is_same<ValT, unsigned long long>::value>
 checkValueRangeImpl(ValT V) {
   static constexpr size_t Limit =
       static_cast<size_t>((std::numeric_limits<int>::max)());
   if (V > Limit)
-    throw sycl::exception(make_error_code(errc::nd_range), NotIntMsg<T>::Msg);
+    throw sycl::exception(make_error_code(errc::nd_range), Msg);
+}
+
+inline void checkMulOverflow(size_t a, size_t b) {
+#ifndef _MSC_VER
+  int Product;
+  // Since we must fit in SIGNED int, we can ignore the upper 32 bits.
+  if (__builtin_mul_overflow(unsigned(a), unsigned(b), &Product)) {
+    throw sycl::exception(make_error_code(errc::nd_range), Msg);
+  }
+#else
+  checkValueRangeImpl(a);
+  checkValueRangeImpl(b);
+  size_t Product = a * b;
+  checkValueRangeImpl(Product);
+#endif
+}
+
+inline void checkMulOverflow(size_t a, size_t b, size_t c) {
+#ifndef _MSC_VER
+  int Product;
+  // Since we must fit in SIGNED int, we can ignore the upper 32 bits.
+  if (__builtin_mul_overflow(unsigned(a), unsigned(b), &Product) ||
+      __builtin_mul_overflow(Product, unsigned(c), &Product)) {
+    throw sycl::exception(make_error_code(errc::nd_range), Msg);
+  }
+#else
+  checkValueRangeImpl(a);
+  checkValueRangeImpl(b);
+  size_t Product = a * b;
+  checkValueRangeImpl(Product);
+
+  checkValueRangeImpl(c);
+  Product *= c;
+  checkValueRangeImpl(Product);
+#endif
+}
+
+// TODO: Remove this function when offsets are removed.
+template <int Dims>
+inline bool hasNonZeroOffset(const sycl::nd_range<Dims> &V) {
+  size_t Product = 1;
+  for (int Dim = 0; Dim < Dims; ++Dim) {
+    Product *= V.get_offset()[Dim];
+  }
+  return (Product != 0);
+}
+#endif
+
+template <int Dims>
+void checkValueRange([[maybe_unused]] const sycl::range<1> &V) {
```

Comment:
How about we merge multiple definitions of `checkValueRange`? Something like the following:
```
template <typename Range>
void checkValueRange([[maybe_unused]] const Range &V) {
#if __SYCL_ID_QUERIES_FIT_IN_INT__
      if constexpr (std::is_same_v<Range, sycl::range<1>>)
            // For 1D range, just check the value against MAX_INT.
           checkValueRangeImpl(V[0]);
     else if constexpr (std::is_same_v<Range, sycl::range<2>>)
           // For 2D range, check if computing the linear range overflows.
          checkMulOverflow(V[0], V[1]);
    else if constexpr (std::is_same_v<Range, sycl::range<3>>)
          // For 3D range, check if computing the linear range overflows.
          checkMulOverflow(V[0], V[1], V[2]);
 #endif
}
```

Reply:
Good idea.  I've done something like this in c50887870d560a3842207e13e008b38bdcc67a35.

I wrote the branches in terms of `Dims`; basically the same, but I thought it might be slightly more readable than comparing the types.

Reply:
Thanks. Looks good.